lldb-dap: Stop using replicated variable ids

Anthony-Eid opened the pull request:
Closes #119784 

Probably closes #147105 as well, but I couldn't test due to #156473: 

This PR fixes two bugs:
1. It generates unique variable reference IDs per suspended debuggee state.
2. It stores all created variables in a stopped state instead of dropping variables in unselected scopes. So it can properly handle all scope/variable requests

It does this by storing all variables in their respective scopes and using that mapping in request handlers that relied on the old mapping. It dynamically creates new variable/scope IDs instead of resetting IDs whenever a new scope is created.

I also removed some unused code as well.

medismailben commented:
@Anthony-Eid could you rebase this PR, that'd be better to look at it

Anthony-Eid committed:
Fix variable request from reusing variable_ids

Anthony-Eid commented:
@medismailben I fixed most of extra lines being added in the PR, but it's not 100% working in the new version. I'll ping you when it's functional.

I believe I'm mainly going to need help rewriting the tests once I get this functionality and following any llvm standards that I'm unaware of.

Anthony-Eid committed:
Add createScope changes

medismailben commented:
Sounds good, let me know when you're ready 🙂

Anthony-Eid committed:
Fix some bugs

Anthony-Eid committed:
Merge branch 'main' into fix-variable-request

Anthony-Eid committed:
remove unused GetScopedsKind func

Anthony-Eid committed:
Fix hardcoded scope references in DAP variable handling

The test and API changes now get scope references dynamically instead of
using hardcoded values.

This fixes the below tests:
- testDAP_memory
- TestDAP_variables
- GetTopLevelScope_ReturnsCorrectScope

Anthony-Eid committed:
Fix set data breakpoint test by using the correct variable reference

Variable reference 1 was hard coded to always be the local scope but
since the var_ref of local scope is dynamic now it uses a helper
function in the test

Anthony-Eid committed:
Have llm (Claude Opus 4) Generate a regression test

Anthony-Eid committed:
Format test files with darker

Anthony-Eid committed:
Remove unused global defines and set First Var Ref to 1

Anthony-Eid commented:
So I ran `/Users/anth/Library/Python/3.9/bin/darker -r HEAD~ ~/Developer/llvm-project/lldb` but it doesn't seem to be formatting the code correctly. What am I missing?

ashgti commented on lldb/tools/lldb-dap/Handler/ScopesRequestHandler.cpp:
```diff
@@ -75,22 +75,31 @@ ScopesRequestHandler::Run(const ScopesArguments &args) const {
+
+  std::vector<Scope> scopes = {};
+
+  int64_t variable_reference = dap.variables.GetNewVariableReference(false);
+  scopes.push_back(CreateScope("Locals", variable_reference,
+                               dap.variables.locals.GetSize(), false));
+
+  dap.variables.AddScopeKind(variable_reference, ScopeKind::Locals, frame_id);
```
Should `AddScopeKind` make a new ref internally and return the ref?

Or when we ready a frame, should this be handled then?

da-viper commented on lldb/tools/lldb-dap/Variables.cpp:
```diff
@@ -103,3 +119,55 @@ lldb::SBValue Variables::FindVariable(uint64_t variablesReference,
+  return true;
+}
+
+void Variables::ReadyFrame(uint32_t frame_id, lldb::SBFrame &frame) {
+  if (m_frames.find(frame_id) == m_frames.end()) {
+
+    auto locals = frame.GetVariables(/*arguments=*/true,
+                                     /*locals=*/true,
+                                     /*statics=*/false,
+                                     /*in_scope_only=*/true);
+
+    auto globals = frame.GetVariables(/*arguments=*/false,
+                                      /*locals=*/false,
+                                      /*statics=*/true,
+                                      /*in_scope_only=*/true);
+
+    auto registers = frame.GetRegisters();
+
+    m_frames.insert(
+        std::make_pair(frame_id, std::make_tuple(locals, globals, registers)));
+  }
+}
```
+1

Anthony-Eid replied:
What exactly is an `SBValueList`?

Would doing something like this work (under m_frames.insert())?

```c
m_variables.insert(locals.var_ref, locals);
m_variables.insert(global.var_ref, global);
m_variables.insert(registers.var_ref, registers);
```

ashgti replied:
A `SBValueList` is the container for the variables, like the local or global or registers.

> Would doing something like this work (under m_frames.insert())?

Yea

Anthony-Eid committed:
Use scopekind instead of hardcoded comparisons in CreateScope

Anthony-Eid commented on lldb/tools/lldb-dap/Handler/ScopesRequestHandler.cpp:
```diff
@@ -41,9 +41,9 @@ static Scope CreateScope(const llvm::StringRef name, int64_t variablesReference,
     scope.presentationHint = Scope::eScopePresentationHintLocals;
-  else if (variablesReference == VARREF_REGS)
+  else if (name == "Registers")
     scope.presentationHint = Scope::eScopePresentationHintRegisters;
```
I passed ScopeKind in to the function instead of name now, and use ScopeKind variant to generate the correct presentation hint and name for the scope. Is that an adequate fix for this?

Anthony-Eid committed:
Stop storing duplicate scopes in variables

I got rid of a variables locals, globals, and register fields and added
some helper methods to access scopes based on a var_ref or a frame_id
and a ScopeKind variant

Anthony-Eid committed:
Apply Python formatting to fix CI darker checks

Anthony-Eid committed:
Hopefully fix failing tests (They're not running on my end)

ashgti commented on lldb/tools/lldb-dap/Variables.cpp:
```diff
@@ -103,3 +120,76 @@ lldb::SBValue Variables::FindVariable(uint64_t variablesReference,
+  }
+}
+
+void Variables::AddScopeKind(int64_t variable_reference, ScopeKind kind,
+                             uint32_t frame_id) {
+
+  m_scope_kinds[variable_reference] = std::make_pair(kind, frame_id);
+}
```
When we ReadyFrame we could go ahead and add the `m_scope_kinds[]` for the frame local/globals/registers. No need to separate those out.

Anthony-Eid replied:
Done

ashgti commented on lldb/tools/lldb-dap/Variables.cpp:
```diff
@@ -103,3 +120,76 @@ lldb::SBValue Variables::FindVariable(uint64_t variablesReference,
+  return nullptr;
+}
+
+void Variables::ReadyFrame(uint32_t frame_id, lldb::SBFrame &frame) {
```
We could have this return a `std::vector<Scope>` for the local/global/register scope.

Anthony-Eid replied:
Done

ashgti commented on lldb/tools/lldb-dap/JSONUtils.cpp:
```diff
@@ -11,6 +11,7 @@
 #include "ExceptionBreakpoint.h"
 #include "LLDBUtils.h"
 #include "ProtocolUtils.h"
+#include "Variables.h"
```
Is variables needed any here? I don't see any usage...

ashgti commented on lldb/tools/lldb-dap/JSONUtils.h:
```diff
@@ -9,6 +9,7 @@
 #ifndef LLDB_TOOLS_LLDB_DAP_JSONUTILS_H
 #define LLDB_TOOLS_LLDB_DAP_JSONUTILS_H
 
+#include "DAP.h"
```
We have DAPForward on line 13, do we need the full DAP here?

Anthony-Eid replied:
I removed it

Anthony-Eid committed:
Actually fix the failing tests for real

Anthony-Eid committed:
Simplify Scopes request handler

Merge ReadyFrame and Switch Frame functions. Also get rid of
AddScopeKind function in favor of inlining it in ReadyFrame

Anthony-Eid committed:
Fix variables test

ashgti commented on lldb/tools/lldb-dap/Variables.h:
```diff
@@ -62,7 +97,11 @@ struct Variables {
+  /// Value = (locals, globals Registers) scopes
+  std::map<uint32_t,
+           std::tuple<lldb::SBValueList, lldb::SBValueList, lldb::SBValueList>>
+      m_frames;
```
I wonder if we could use a strategy like encoding the `(thread << 19) + (frame id << 2) + <scope offset>` where the scope offset is `locals = 0, globals = 1  registers = 2`.

That would effectively mean the scopes reference variable would always be formatted like:

```
[ -  variable reference bits - ]
[32 ---------- 16 ----------- 0]
TTTTTTTTTTTTTFFFFFFFFFFFFFFFFFSS
```

Where `T` is the thread id, `F` is the frame id and `S` is the scope id.

Anthony-Eid replied:
Would you still want to do something similar here now that we're using the dap_frame_id as the key?

ashgti commented on lldb/tools/lldb-dap/Variables.h:
```diff
@@ -62,7 +97,11 @@ struct Variables {
-  int64_t m_next_temporary_var_ref{VARREF_FIRST_VAR_IDX};
+  /// Key = frame_id
+  /// Value = (locals, globals Registers) scopes
+  std::map<uint32_t,
```
I just realized that the frame id isn't unique enough for a key here.

The frame_id is unique to the thread it comes from but not across threads, e.g. 
```
$ lldb -p 111111
(lldb) script
>>> lldb.process.threads[0].frames[0].GetFrameID()
0
>>> lldb.process.threads[1].frames[0].GetFrameID()
0
```

Anthony-Eid replied:
Sorry it took me so long to get back to this. I just fixed this by using the stack frame id field from the DAP representation. 

https://github.com/Anthony-Eid/llvm-project/blob/c4155cc279323f5014dba9e4495ec4bef71e0a31/lldb/tools/lldb-dap/Handler/ScopesRequestHandler.cpp#L39-L40

Please let me know if there's something I'm missing with this solution

Anthony-Eid committed:
Simplify ScopeData and improve naming in ready frame function

Anthony-Eid committed:
Switch Variables.m_frames key to use thread_id and frame_id as key

This addresses @ashgti feedback by making the frame key unique across
threads.

Anthony-Eid committed:
Merge origin/main into fix-variable-request

Resolve conflicts by:
- Keeping dynamic scope reference lookup in lldbdap_testcase.py
- set_local() and set_global() now use set_variable() to properly
  trigger invalidated and memory events
- Test files use set_local() which internally uses dynamic scope lookup

Anthony-Eid committed:
Add test_variable_id_uniqueness_simple regression test

Anthony-Eid commented:
@medismailben Sorry for the delay, this PR should be ready for review again!

medismailben committed:
Merge branch 'main' into fix-variable-request

Anthony-Eid committed:
Fix duplicate function definition

Anthony-Eid committed:
Move m_next_temporary_var_ref next to permanent var ref

Anthony-Eid committed:
Add e prefix to scopes kind enum

Anthony-Eid committed:
Fix comment copy and paste error

Anthony-Eid committed:
Create a FrameScope object to store a stack frames scopes

I also made a helper method to get the scope kind based on the
eScopeKind enum

Anthony-Eid committed:
Switch VARREF_FIRST_VAR_IDX from a #define variable to a constexpr

This follows what `PermanentVariableStartIndex` does and is more inline
with newer c++ standards.

I renamed the variable to `TemporaryVariableStartIndex` as well

Anthony-Eid commented on lldb/tools/lldb-dap/Variables.h:
```diff
@@ -9,22 +9,43 @@
-#define VARREF_LOCALS (int64_t)1
-#define VARREF_GLOBALS (int64_t)2
-#define VARREF_REGS (int64_t)3
+#define VARREF_FIRST_VAR_IDX (int64_t)1
```
Moved this to a `constexpr` named `TemporaryVariableStartIndex`

Anthony-Eid committed:
Add scope lookup helper for DAP VariablesTest

Anthony-Eid committed:
Merge remote-tracking branch 'origin' into fix-variable-request

Anthony-Eid committed:
Explicitly add optional include to variable.h

ashgti commented on lldb/tools/lldb-dap/Variables.h:
```diff
@@ -62,7 +95,11 @@ struct Variables {
+  /// Key = dap_frame_id (encodes both thread index ID and frame ID)
+  /// Value = (locals, globals, registers) scopes
+  std::map<uint64_t,
+           std::tuple<lldb::SBValueList, lldb::SBValueList, lldb::SBValueList>>
```
You can add a `llvm_unreachable("unknown scope kind");` after the switch statement.

ashgti commented on lldb/tools/lldb-dap/Variables.cpp:
```diff
@@ -8,27 +8,72 @@
   m_referencedvariables.clear();
+  m_scope_kinds.clear();
+  m_frames.clear();
+  m_next_temporary_var_ref = TemporaryVariableStartIndex;
```
I think we shouldn't need to reset this value on clear, otherwise we'd end up reusing ids again.

Anthony-Eid replied:
I can remove the reset, but it's valid to do whenever a stop state resumes execution which is when clear get's called to my knowledge.

JDevlieghere reviewed (commented):
A few stylistic nits.

Anthony-Eid committed:
Update lldb/tools/lldb-dap/Variables.cpp

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

Anthony-Eid committed:
Update lldb/tools/lldb-dap/Variables.cpp

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

Anthony-Eid committed:
Update lldb/tools/lldb-dap/Variables.cpp

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

Anthony-Eid committed:
Add llvm unreachable

Anthony-Eid committed:
Update lldb/tools/lldb-dap/Variables.cpp

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

Anthony-Eid committed:
Move scope field initialization closer to func args

Anthony-Eid commented on lldb/tools/lldb-dap/Variables.h:
```diff
@@ -47,13 +91,27 @@ struct Variables {
   lldb::SBValue FindVariable(uint64_t variablesReference, llvm::StringRef name);
 
+  /// Initialize a frame if it hasn't been already, otherwise do nothing
+  std::vector<protocol::Scope> ReadyFrame(const uint64_t dap_frame_id,
```
How do you feel about initialize Frame?

da-viper replied:
CreateScopes or MakeScopes as that is what we are using the sbframe to do in the function.

da-viper commented on lldb/tools/lldb-dap/Variables.cpp:
```diff
@@ -103,3 +146,89 @@ lldb::SBValue Variables::FindVariable(uint64_t variablesReference,
+  scopes.push_back(CreateScope(
+      eScopeKind::Globals, globals_ref,
+      GetScope(dap_frame_id, eScopeKind::Globals)->GetSize(), false));
+  m_scope_kinds[globals_ref] =
```
Could we change how the `readyFrame` function works as we are always checking the map every time we need a `FrameScope` instead do it once and use that value. something like.

```cpp
std::vector<protocol::Scope> Variables::ReadyFrame(const uint64_t dap_frame_id,
                                                   lldb::SBFrame &frame) {
  auto iter = m_frames.find(dap_frame_id);
  if (iter == m_frames.end()) {
    auto locals = frame.GetVariables(/*arguments=*/true,
                                     /*locals=*/true,
                                     /*statics=*/false,
                                     /*in_scope_only=*/true);

    auto globals = frame.GetVariables(/*arguments=*/false,
                                      /*locals=*/false,
                                      /*statics=*/true,
                                      /*in_scope_only=*/true);

    auto registers = frame.GetRegisters();

    iter = m_frames.emplace(dap_frame_id, FrameScopes{locals, globals, registers})
            .first;
  }

  const FrameScopes &frame_scopes = iter->second;
  auto create_and_store_scope = [&](eScopeKind kind, uint32_t size) {
    const int64_t ref = GetNewVariableReference(false);
    m_scope_kinds.try_emplace(ref, kind, dap_frame_id);
    return CreateScope(kind, ref, size, false);
  };

  return {
      create_and_store_scope(eScopeKind::Locals, frame_scopes.locals.GetSize()),
      create_and_store_scope(eScopeKind::Globals,
                             frame_scopes.globals.GetSize()),
      create_and_store_scope(eScopeKind::Registers,
                             frame_scopes.registers.GetSize())};
}
```

Anthony-Eid committed:
Update lldb/tools/lldb-dap/Variables.cpp

Co-authored-by: Ebuka Ezike <yerimyah1@gmail.com>

Anthony-Eid committed:
Update lldb/tools/lldb-dap/Variables.cpp

Co-authored-by: Ebuka Ezike <yerimyah1@gmail.com>

da-viper commented:
Hello @Anthony-Eid, It seems it's all complete except the last changes. You can ping me if there is an issue. To avoid rebasing as the code base is changing quite a lot.

Anthony-Eid committed:
Rename ScopeKind and add unsigned typing

Anthony-Eid committed:
merge with upstream

Anthony-Eid committed:
Merge branch 'main' into fix-variable-request

Anthony-Eid committed:
rename ReadyFrame to CreateScopes

Anthony-Eid committed:
Make GetTopLevelScope return an optional<ScopeData>

Anthony-Eid committed:
Clean up CreateScope function

Anthony-Eid committed:
Fix variables test

Anthony-Eid commented:
> > we're so close
> 
> @Anthony-Eid You need a merge ?

Do you mean merging this PR? If so yes please!

I merged the PR with origin/main last night so I don't think that's needed

medismailben commented:
@Anthony-Eid keep an eye on this PR and on the bots on https://lab.llvm.org/buildbot/#/builders since it tests more configuration that PR testing.

da-viper commented:
/cherry-pick 6ca8f79fc88971412e6e5692ea86d680cd438fe3
